### PR TITLE
Note function entry points.

### PR DESCRIPTION
**Pull request by chriseth**
References https://github.com/ethereum/solidity/issues/6485 and #9332

### CODE REVIEW

**Comment by chriseth** [comment]
This could also output the starting PCs and names of the yul functions.

**Comment by chriseth** [comment]
The output looks like this now: https://gist.github.com/chriseth/540c55016ae021e75d77d04bbea5adfc

Which other information would you like?

@yann300 @alcuadrado @iamdefinitelyahuman @haltman-at

**Comment by chriseth** [comment]
@mrchico I'm not sure I can satisfy your request to mark the modifier entry points, but here are the function entry points.

**Comment by chriseth** [comment]
One thing to note is that if the optimizer joins two functions then one of the functions will be ignored.

**Comment by haltman-at** [comment]
Interesting!  So this is how you get around the problem of it being dependent on contract, huh?  By making it a per-contract output.

Despite my earlier comment (sorry) I'm not sure how much I'd actually use this -- after all, Truffle Debugger already does a bunch of work to determine function entry points.  Although, I guess we could use this if available?  Hm... @gnidan suggested that maybe this could also be more reliable if optimization was turned on?  Not something I've tried.  Optimization will mess with a lot of things one way or another.

The talk of modifier entry points is quite interesting, although I'm pretty confused as to how that would work given that modifiers don't have a single location at the moment.

I guess I can think of two possible uses for this, for my own purposes:

1. Decoding internal function pointers.  For that purpose it might be nice to have the entry point in constructor context as well, but I guess one can do without that, seeing as the pointer will always contain the deployed entry point.  This is the purpose for which we already determine function entry points.

2. Telling when we're entering a function -- OK, that's not something we really need new information to be able to do; we already have jump markings for that.  (Note for this purpose we *would* also need the entry points in constructor context, to tell when we're entering them while doing a deployment.)  But this *could* potentially be useful for things that *don't* get jump markings -- namely, modifiers, as mentioned above.  I've previously talked about changing modifier depth to make it more usable for this purpose, but maybe the real solution is not to use modifier depth at all, but just to have a direct record of the entry points!

...except, of course, for the problem that modifiers don't actually work here due to not having a single entry point.  Um, maybe there could be a way to list multiple entry points...?

This also could be useful for telling when we're entering things other than named functions -- i.e., constructors, fallback functions, and receive functions.  Those at least should have single entry points!  The most useful one here is constructors, for the same reason I want it for modifiers -- what I care about is not so much when we enter the main constructor, but rather when we might be entering any base constructors.  The main constructor itself, as well as fallback/receive, would just be kind of nice for printing stack traces, I guess.  I guess that's sort of a third use.  Except that, with modifiers or base constructors, using it for those purposes would also require a way to tell when we're *exiting* so things don't get unbalanced...

OK, maybe I shouldn't try to incorporate it into that, and you should ignore the things that would only be useful for that...

**Comment by MrChico** [comment]
This looks great! Not sure what the `TODO` in the cli opts is referring to exactly, but I hope this stays available to the solc cli and not just json opts.

**Comment by chriseth** [comment]
@haltman-at good news: Of course this is available both for the construction and the runtime context. Bad news: For both constructor and modifier the entry point is not a jumpdest and thus rather difficult to implement.

**Comment by chriseth** [comment]
It might make sense to include local variables as outlined in this document: https://www.notion.so/Local-variables-symbols-Initial-proposal-7a317eb434204d43b7dc86e250700400

**Comment by haltman-at** [comment]
> @haltman-at good news: Of course this is available both for the construction and the runtime context. Bad news: For both constructor and modifier the entry point is not a jumpdest and thus rather difficult to implement.

Ah, I was afraid that would probably be the case, but figured I'd ask!

**Comment by chriseth** [comment]
If we add local variable scopes (start init, end init, end scope) to this information, it could be done by using assembly items that result in zero bytes in the output and only contain debugging information.

**Comment by chriseth** [comment]
Local variable poc implementation: https://github.com/ethereum/solidity/compare/develop...jcaracciolo:debugging-symbols

File: libyul/backends/evm/AbstractAssembly.h [hunk 1]
```diff
@@ -70,7 +70,7 @@ class AbstractAssembly
 	/// Generate a new unique label.
 	virtual LabelID newLabelId() = 0;
 	/// Returns a label identified by the given name. Creates it if it does not yet exist.
-	virtual LabelID namedLabel(std::string const& _name) = 0;
+	virtual LabelID namedLabel(std::string const& _name, size_t _params, size_t _returns) = 0;
```
Comment (cameel) [inline libyul/backends/evm/AbstractAssembly.h]: I think that the docstring should mention what happens to `_params` and `_returns`. E.g. is it considered a different label if the name is the same but those two are different?

File: libsolidity/interface/CompilerStack.h [hunk 2]
```diff
@@ -299,6 +299,9 @@ class CompilerStack: boost::noncopyable
 	/// Prerequisite: Successful call to parse or compile.
 	Json::Value const& storageLayout(std::string const& _contractName) const;
 
+	Json::Value const& functionDebugData(std::string const& _contractName) const;
+	Json::Value const& runtimeFunctionDebugData(std::string const& _contractName) const;
```
Comment (cameel) [inline libsolidity/interface/CompilerStack.h]: Naming should be updated to match #10934.

File: libsolidity/interface/CompilerStack.cpp [hunk 3]
```diff
@@ -751,6 +769,26 @@ Json::Value const& CompilerStack::storageLayout(Contract const& _contract) const
...
+	evmasm::LinkerObject const& object = _runtime ? _contract.runtimeObject : _contract.object;
+	solAssert(_contract.compiler, "");
+	Json::Value data = Json::arrayValue;
+	for (auto const& [name, info]: object.functionEntryPointInfo)
+	{
+		Json::Value item = Json::objectValue;
+		if (Declaration const* declaration = _contract.compiler->declarationOfFunctionEntryLabel(name, _runtime))
+			item["id"] = declaration->id();
+		else
+			item["internal_name"] = name;
```
Comment (cameel) [inline libsolidity/interface/CompilerStack.cpp]:
```suggestion
			item["internalName"] = name;
```

File: solc/CommandLineInterface.cpp [hunk 4]
```diff
@@ -453,6 +453,9 @@ void CommandLineInterface::handleStorageLayout(string const& _contract)
 	if (!m_args.count(g_argStorageLayout))
 		return;
 
+	// TODO remove after testing
+	sout() << "Function debug data: " << endl << jsonPrettyPrint(m_compiler->runtimeFunctionDebugData(_contract)) << endl;
```
Comment (cameel) [inline solc/CommandLineInterface.cpp]: Do we need a new CLI option to output the debug data or should it be included somewhere in the existing output? I don't see this PR including it anywhere apart from this debug print.

File: libsolidity/codegen/CompilerContext.h [hunk 5]
```diff
@@ -211,7 +215,10 @@ class CompilerContext
 	/// @returns a new tag without pushing any opcodes or data
 	evmasm::AssemblyItem newTag() { return m_asm->newTag(); }
 	/// @returns a new tag identified by name.
-	evmasm::AssemblyItem namedTag(std::string const& _name) { return m_asm->namedTag(_name); }
+	evmasm::AssemblyItem namedTag(std::string const& _name, size_t _params, size_t _returns)
```
Comment (cameel) [inline libsolidity/codegen/CompilerContext.h]: The docstring should describe `_params` and `_returns` too.

File: libsolidity/interface/CompilerStack.cpp (hunk 3 above)
Comment (cameel) [inline libsolidity/interface/CompilerStack.cpp]: In what kind of situation we can get a named tag in the assembly along with the numbers of parameters but without a declaration associated with it?

EDIT: Looking at the generated output - so it's meant mostly for functions generated by the compiler? I think it would be nice to have a comment explaining that here.

**Comment by cameel** [comment]
@chriseth Could you summarize what is left to do here? This PR has been hanging here for a long time now so I think the best course of action would be to keep the functionality to a minimum here (i.e. deal only with the function entry points) to get it merged and move the discussion to an issue where we can continue discussing other ideas (entry points for constructors and modifiers; variable scopes; adding more debug info) and implement them in subsequent PRs.

After reviewing this, I think that for function entry points we only need to make the info available via CLI and Standard JSON and add the minor tweaks from my review comments. This also needs a rebase which will unfortunately will not be a fast-forward - `InvertibleMap` has been removed and there were some changes around this code, like the naming change from #10934.

What do you think?

**Comment by chriseth** [comment]
Updated with a slightly different approach that hopefully still works when compiling via yul.

Left to do:
 - document changes to standard json (especially the possible null values)
 - add tests
 - decide if we want to support this from cmdline or not.

**Comment by cameel** [comment]
> decide if we want to support this from cmdline or not.

I think it would be best to keep parity between Standard JSON and CLI if possible.

**Comment by chriseth** [comment]
Thinsg like sourcemap are just in "combined json", so I will only add it to combined json.

File: Changelog.md [hunk 6]
```diff
@@ -6,6 +6,7 @@ Language Features:
 
 
 Compiler Features:
+ * Standard JSON / compact JSON: New artefact "functionDebugData" that contains bytecode offsets of entry points of functions and potentially more information in the future.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Standard JSON / combined JSON: New artefact "functionDebugData" that contains bytecode offsets of entry points of functions and potentially more information in the future.
```

File: Changelog.md [hunk 7]
```diff
@@ -6,6 +6,7 @@ Language Features:
 
 
 Compiler Features:
+ * Standard JSON / combined JSON: New artefact "functionDebugData" that contains bytecode offsets of entry points of functions and potentially more information in the future.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Standard JSON / combined JSON: New artifact "functionDebugData" that contains bytecode offsets of entry points of functions and potentially more information in the future.
```

File: docs/using-the-compiler.rst [hunk 8]
```diff
@@ -476,6 +477,17 @@ Output Description
               "legacyAssembly": {},
               // Bytecode and related details.
               "bytecode": {
+                // Debugging data at the level of functions.
+                "functionDebugData": {
+                  // Now follows a set of functions including compiler-internal and
+                  // user-defined function. The set does not have to be complete.
+                  "@mint_13": { // Internal name of the function
+                    "entryPoint": 128, // Byte offset into the code where the function starts
```
Comment (cameel) [inline docs/using-the-compiler.rst]: I'd make it more obvious that this marks a location in the binary, not in the source code.
```suggestion
                    "entryPoint": 128, // Byte offset into the bytecode where the function starts
```

File: test/cmdlineTests/function_debug_info/input.sol [hunk 9]
```diff
@@ -0,0 +1,6 @@
+contract C {
+  function f(uint[] calldata x) pure external returns (uint) { return x[0]; }
+  // This will be optimized out
+  function g() pure internal {}
+  mapping(uint => uint) public t;
```
Comment (cameel) [inline test/cmdlineTests/function_debug_info/input.sol]:
```suggestion
    function f(uint[] calldata x) pure external returns (uint) { return x[0]; }
    // This will be optimized out
    function g() pure internal {}
    mapping(uint => uint) public t;
```

Same in the other test.

File: test/cmdlineTests/function_debug_info/exit [hunk 10]
```diff
@@ -0,0 +1 @@
+0
```
Comment (cameel) [inline test/cmdlineTests/function_debug_info/exit]: This file is optional when return code is 0. I'd remove it to decrease the number of files here.

Same in the other test.

File: test/cmdlineTests/function_debug_info/input.sol [hunk 11]
```diff
@@ -0,0 +1,6 @@
+contract C {
```
Comment (cameel) [inline test/cmdlineTests/function_debug_info/input.sol:3]: I'd add license and version pragma to silence the warnings. They're just noise here.
```suggestion
// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.0;

contract C {
```

Same in the other test.

File: docs/using-the-compiler.rst (hunk 8 above)
Comment (cameel) [inline docs/using-the-compiler.rst]: I see that `entryPoint` is sometimes omitted. This comment should explain when it can happen.

It's when `m_tagPositionsInBytecode` contains `numeric_limits<size_t>::max()` for the tag. I.e. there was no code generated for the function? Is it the case of a function being optimized out/inlined?

File: test/cmdlineTests/function_debug_info/output [hunk 12]
```diff
@@ -0,0 +1,43 @@
+{
+  "contracts":
+  {
+    "function_debug_info/input.sol:C":
+    {
+      "function-debug": {},
```
Comment (cameel) [inline test/cmdlineTests/function_debug_info/output]: Maybe you could add a constructor in the input and do some calls from it to make sure this works too?

File: test/cmdlineTests/function_debug_info_via_yul/output [hunk 13]
```diff
@@ -0,0 +1,11 @@
+{
+  "contracts":
+  {
+    "function_debug_info_via_yul/input.sol:C":
+    {
+      "function-debug": {},
+      "function-debug-runtime": {}
```
Comment (cameel) [inline test/cmdlineTests/function_debug_info_via_yul/output:7]: So it does not work for EVM bytecode generated from Yul? What do we need to make this happen?

File: test/cmdlineTests/standard_function_debug_info/input.json [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  "outputSelection": {
+   "*": {
+    "A": [
+     "evm.deployedBytecode.funtionDebugData",
+     "evm.bytecode.functionDebugData"
+    ]
+   }
+  }
+ }
+}
```
Comment (cameel) [inline test/cmdlineTests/standard_function_debug_info/input.json]: Single-space indents really hurt to look at :P
```suggestion
{
    "language": "Solidity",
    "sources": {
        "a.sol": {
            "content": "// SPDX-License-Identifier: GPL-3.0\npragma abicoder v2; contract A { function f(uint[] memory) public view returns (uint256) { } }"
        }
    },
    "settings": {
        "evmVersion": "petersburg",
        "outputSelection": {
            "*": {
                "A": [
                    "evm.deployedBytecode.funtionDebugData",
                    "evm.bytecode.functionDebugData"
                ]
            }
        }
    }
}

```

File: test/cmdlineTests/standard_function_debug_info/input.json [hunk 15]
```diff
@@ -0,0 +1,19 @@
+{
+ "language": "Solidity",
+ "sources": {
+  "a.sol": {
+   "content": "// SPDX-License-Identifier: GPL-3.0\npragma abicoder v2; contract A { function f(uint[] memory) public view returns (uint256) { } }"
```
Comment (cameel) [inline test/cmdlineTests/standard_function_debug_info/input.json]:
```suggestion
   "content": "// SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0; pragma abicoder v2; contract A { function f(uint[] memory) public view returns (uint256) { } }"
```

File: test/cmdlineTests/standard_function_debug_info/output.json [hunk 16]
```diff
@@ -0,0 +1,4 @@
+{"contracts":{"a.sol":{"A":{"evm":{"bytecode":{"functionDebugData":{}}}}}},"errors":[{"component":"general","errorCode":"3420","formattedMessage":"Warning: Source file does not specify required compiler version!
+--> a.sol
```
Comment (cameel) [inline test/cmdlineTests/standard_function_debug_info/output.json]: I just realized that I could have used `--pretty-json` flag to make my `--combined-json` test output in tests look readable. I think we should make this flag work for Standard JSON too. Or is that already possible?

Not something to add in this PR of course.

File: test/cmdlineTests/standard_function_debug_info/input.json [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+  }
+ },
+ "settings": {
+  "evmVersion": "petersburg",
```
Comment (cameel) [inline test/cmdlineTests/standard_function_debug_info/input.json]: Why petersburg? It was just in the test case you used as a base and is not really necessary here, right?

File: test/cmdlineTests/standard_function_debug_info/output.json [hunk 18]
```diff
@@ -0,0 +1,4 @@
+{"contracts":{"a.sol":{"A":{"evm":{"bytecode":{"functionDebugData":{}}}}}},"errors":[{"component":"general","errorCode":"3420","formattedMessage":"Warning: Source file does not specify required compiler version!
```
Comment (cameel) [inline test/cmdlineTests/standard_function_debug_info/output.json]: Something seems to be broken. `functionDebugData` is empty here.

File: test/cmdlineTests/function_debug_info_via_yul/output (hunk 13 above)
Comment (chriseth) [reply test/cmdlineTests/function_debug_info_via_yul/output:7]: Exactly. We need to extend the debug info data structure that @mijovic is currently working on and pull it through using annotations.

File: test/cmdlineTests/standard_function_debug_info/output.json (hunk 16 above)
Comment (chriseth) [reply test/cmdlineTests/standard_function_debug_info/output.json]: It is useful for combined-json because the output is not a single file. For standard-json you can just use `| jq`.

**Comment by chriseth** [comment]
This seems to be broken for standard-json. need to investigate.

File: test/cmdlineTests/standard_function_debug_info/output.json (hunk 16 above)
Comment (cameel) [reply test/cmdlineTests/standard_function_debug_info/output.json]: True. You cannot use it in command-line tests though.

**Review by cameel — APPROVED** [review]
I think that the newly added parameters should be documented, at least in `AbstractAssembly.h`. Especially the fact that they refer to the number of slots rather than parameters.

Other than that, everything seems to be fine and the JSON test is now fixed so I'm approving.
